modules/javafx.base/src/main/java/com/sun/javafx/collections/ConcatenatedObservableList.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.sun.javafx.collections;
+
+import javafx.collections.ListChangeListener;
+import javafx.collections.ObservableList;
+import javafx.collections.ObservableListBase;
+import javafx.collections.WeakListChangeListener;
+import java.util.Objects;
+
+public final class ConcatenatedObservableList<E> extends ObservableListBase<E> {
+
+    private final ObservableList<? extends E>[] lists;
+    private final ListChangeListener<? super E> listChangeListener;
+    private int size;
+
+    public ConcatenatedObservableList(ObservableList<? extends E>[] lists) {
+        this.lists = lists;
+        this.listChangeListener = this::onListChanged;
+
+        ListChangeListener<? super E> weakListChangeListener = new WeakListChangeListener<>(listChangeListener);
+        for (ObservableList<? extends E> list : lists) {
+            list.addListener(weakListChangeListener);
+            size = addSize(list.size());
+        }
+    }
+
+    private int computeListOffset(ObservableList<? extends E> list) {
+        int offset = 0;
+
+        for (ObservableList<? extends E> l : lists) {
+            if (l != list) {
+                offset += l.size();
+            } else {
+                return offset;
+            }
+        }
+
+        throw new IllegalArgumentException("list");
+    }
+
+    private void onListChanged(ListChangeListener.Change<? extends E> change) {
+        int listOffset = computeListOffset(change.getList());
+
+        while (change.next()) {
+            beginChange();
```

Comment:
I've added documentation that the list is in an undefined state if the sum of its constituent elements exceeds `Integer.MAX_VALUE`. As an alternative, I also thought about tracking the excess size with a `long` size field. However, while we may avoid breaking this list instance, we still potentially break listeners since change events will be incorrect.